flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeUploadScheduler.java

```diff
@@ -0,0 +1,158 @@
+                uploader.upload(singletonList(uploadTask)).complete();
+            }
+
+            @Override
+            public void close() throws Exception {
+                uploader.close();
+            }
+        };
+    }
+
+    static StateChangeUploadScheduler fromConfig(
+            ReadableConfig config, ChangelogStorageMetricGroup metricGroup) throws IOException {
+        Path basePath = new Path(config.get(BASE_PATH));
+        long bytes = config.get(UPLOAD_BUFFER_SIZE).getBytes();
+        checkArgument(bytes <= Integer.MAX_VALUE);
+        int bufferSize = (int) bytes;
+        StateChangeFsUploader store =
+                new StateChangeFsUploader(
+                        basePath,
+                        basePath.getFileSystem(),
+                        config.get(COMPRESSION_ENABLED),
+                        bufferSize,
+                        metricGroup);
+        BatchingStateChangeUploadScheduler batchingStore =
+                new BatchingStateChangeUploadScheduler(
+                        config.get(PERSIST_DELAY).toMillis(),
+                        config.get(PERSIST_SIZE_THRESHOLD).getBytes(),
+                        RetryPolicy.fromConfig(config),
+                        store,
+                        config.get(NUM_UPLOAD_THREADS),
+                        config.get(IN_FLIGHT_DATA_LIMIT).getBytes(),
+                        metricGroup);
+        return batchingStore;
+    }
+
+    default AvailabilityProvider getAvailabilityProvider() {
+        return () -> AvailabilityProvider.AVAILABLE;
+    }
+
+    @ThreadSafe
+    final class UploadTask {
+        final Collection<StateChangeSet> changeSets;
+        final BiConsumer<List<SequenceNumber>, Throwable> failureCallback;
+        final Consumer<List<UploadResult>> successCallback;
+        final AtomicBoolean finished = new AtomicBoolean();
+
+        public UploadTask(
+                Collection<StateChangeSet> changeSets,
+                Consumer<List<UploadResult>> successCallback,
+                BiConsumer<List<SequenceNumber>, Throwable> failureCallback) {
+            this.changeSets = new ArrayList<>(changeSets);
+            this.failureCallback = failureCallback;
+            this.successCallback = successCallback;
+        }
+
+        public void complete(List<UploadResult> results) {
+            if (finished.compareAndSet(false, true)) {
+                successCallback.accept(results);
+            }
+        }
+
+        public void fail(Throwable error) {
+            if (finished.compareAndSet(false, true)) {
+                failureCallback.accept(
+                        changeSets.stream()
+                                .map(StateChangeSet::getSequenceNumber)
+                                .collect(toList()),
+                        error);
+            }
+        }
+
+        public long getSize() {
+            long size = 0;
+            for (StateChangeSet set : changeSets) {
+                size = set.getSize();
```

Comment:
It should be `size += set.getSize();`

Reply:
I think we can make the size of `changeSets` larger than 1, or avoid the `singletonList` in the `List<StateChangeSet> getChanges(int size)` method within `BatchingStateChangeUploadSchedulerTest` to find this bug.